src/Microsoft.Diagnostics.EventFlow.Inputs.Trace/TraceInput.cs

```diff
@@ -161,15 +162,35 @@ public IDisposable Subscribe(IObserver<EventData> observer)
 
         protected override void Dispose(bool disposing)
         {
+            if (!this.disposed)
+            {
+                lock (this.subject)
+                {
+                    if (!this.disposed)
+                    {
+                        try
+                        {
+                            this.disposed = true;
+                            Trace.Listeners.Remove(this);
+                            this.subject.Dispose();
+                        }
+                        finally
+                        {
+                            this.disposed = true;
+                        }
+                    }
+                }
+            }
+
             base.Dispose(disposing);
```

Comment:
According to .NET design guidelines one should be able to call Dispose() multiple times on the same object, with no ill effects. In this particular piece of code we want to ensure that we call Trace.Listeners.Remove(this) _exactly once_ as part of disposing, regardless how many times Dispose() is called